src/ray/core_worker/experimental_mutable_object_provider.h

```diff
@@ -0,0 +1,110 @@
+#include "ray/core_worker/common.h"
+#include "ray/core_worker/experimental_mutable_object_manager.h"
+#include "ray/core_worker/store_provider/plasma_store_provider.h"
+#include "ray/raylet_client/raylet_client.h"
+#include "ray/rpc/client_call.h"
+#include "ray/rpc/worker/core_worker_client_pool.h"
+
+namespace ray {
+namespace core {
+namespace experimental {
+
+// This class coordinates the transfer of mutable objects between different nodes. It
+// handles mutable objects that are received from remote nodes, and it also observes local
+// mutable objects and pushes them to remote nodes as needed.
+class MutableObjectProvider {
+ public:
+  typedef std::function<std::shared_ptr<MutableObjectReaderInterface>(
+      const NodeID &node_id)>
+      RayletFactory;
+
+  MutableObjectProvider(std::shared_ptr<plasma::PlasmaClientInterface> plasma,
+                        RayletFactory factory);
+
+  ~MutableObjectProvider();
+
+  std::unique_ptr<rpc::ClientCallManager> &client_call_manager() {
+    return client_call_manager_;
+  }
+
+  ray::experimental::MutableObjectManager &object_manager() { return object_manager_; }
+
+  /// Registers a writer channel for `object_id` on this node. On each write to this
+  /// channel, the write will be sent via RPC to node `node_id`.
+  /// \param[in] object_id The ID of the object.
+  /// \param[in] node_id The ID of the node to write to.
+  void RegisterWriterChannel(const ObjectID &object_id, const NodeID &node_id);
+
+  void HandleRegisterMutableObject(const ObjectID &object_id,
+                                   int64_t num_readers,
+                                   const ObjectID &local_object_id);
+
+  /// RPC callback for when a writer pushes a mutable object over the network to a reader
+  /// on this node.
+  void HandlePushMutableObject(const rpc::PushMutableObjectRequest &request,
+                               rpc::PushMutableObjectReply *reply);
+
+ private:
+  struct LocalInfo {
+    int64_t num_readers;
+    ObjectID local_object_id;
+  };
+
+  // Registers a reader channel for `object_id` on this node.
+  void RegisterReaderChannel(const ObjectID &object_id);
+
+  // Listens for local changes to `object_id` and sends the changes to remote nodes via
+  // the network.
+  void PollWriterClosure(const ObjectID &object_id,
+                         std::shared_ptr<MutableObjectReaderInterface> reader);
+
+  // Kicks off `io_service_`.
+  void RunIOService();
+
+  // The plasma store.
+  std::shared_ptr<plasma::PlasmaClientInterface> plasma_;
+
+  // Object manager for the mutable objects.
+  ray::experimental::MutableObjectManager object_manager_;
+
+  // Protects `cross_node_map_`.
+  absl::Mutex cross_node_map_lock_;
+  // Maps the remote node object ID (i.e., the object ID that the remote node writes to)
+  // to the corresponding local object ID (i.e., the object ID that the local node reads
+  // from) and the number of readers.
+  std::unordered_map<ObjectID, LocalInfo> cross_node_map_;
```

Comment:
It's a bit unclear from the name if this map is used by the writer or the reader. Maybe: `remote_writer_object_to_local_reader_`?